### PR TITLE
NFT pinning optimisation

### PR DESCRIPTION
**Pull request by cypt4**
Resolves https://github.com/brave/brave-browser/issues/29230
Before: ipfs://bafy/1/1.png is given and ipfs://bafy is pinned recursively.
After: bafy and bafy/1 pinned non-recursively and bafy/1/1.png pinned recursively.

Resolves 

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
0) Overall test that NFT pinning works.
1) Pin some NFT which has metadata or image format like ipfs.io/ipfs/CID/../*.jpg(json,png etc).
2) Check whether CID is pinned using direct mode by calling curl -X POST http://127.0.0.1:45001/api/v0/pin/ls?arg=CID
3) Check whether image\metadata of the NFT are pinned recursively by calling
http://127.0.0.1:45001/api/v0/pin/ls?arg=%2Fipfs%2F<CID>%2F<PATH>

### CODE REVIEW

File: components/ipfs/pin/ipfs_local_pin_service.cc [hunk 1]
```diff
@@ -11,61 +11,161 @@
...
 #include "components/prefs/scoped_user_pref_update.h"
 
 namespace ipfs {
 
+bool PinData::operator==(const PinData& d) const {
+  return this->cid == d.cid && this->pinning_mode == d.pinning_mode;
+}
+bool PinData::operator==(const PinData& d) {
+  return this->cid == d.cid && this->pinning_mode == d.pinning_mode;
+}
```
Comment (supermassive) [inline components/ipfs/pin/ipfs_local_pin_service.cc:28]: Why we need non-const version?

File: components/ipfs/pin/ipfs_local_pin_service.cc [hunk 2]
```diff
@@ -11,61 +11,161 @@
...
+    builder += part;
+    bool recursive = ++counter == splitted.size();
+    result.push_back(PinData{
+        builder, recursive ? PinningMode::RECURSIVE : PinningMode::DIRECT});
+  }
+  return result;
+}
+
+absl::optional<std::vector<ipfs::PinData>>
+IpfsLocalPinService::ExtractMergedPinData(std::vector<std::string> ipfs_urls) {
```
Comment (supermassive) [inline components/ipfs/pin/ipfs_local_pin_service.cc:76]: const &

File: components/ipfs/pin/ipfs_local_pin_service.cc [hunk 3]
```diff
@@ -11,61 +11,161 @@
...
+  std::vector<ipfs::PinData> result;
+  for (const auto& ipfs_url : ipfs_urls) {
+    auto pins_data_for_url = ExtractPinData(ipfs_url);
+    if (!pins_data_for_url) {
+      return absl::nullopt;
+    }
+    for (const auto& item : pins_data_for_url.value()) {
+      if (std::find(result.begin(), result.end(), item) == result.end()) {
```
Comment (supermassive) [inline components/ipfs/pin/ipfs_local_pin_service.cc:84]: base::Contains

File: components/ipfs/pin/ipfs_local_pin_service.cc [hunk 4]
```diff
@@ -11,61 +11,154 @@
...
+    return absl::nullopt;
+  }
+  std::vector<PinData> result;
+  std::string builder = "/ipfs";
+  for (const auto& part : slit) {
+    builder += "/";
+    builder += part;
+    result.push_back(PinData{builder, PinningMode::DIRECT});
+  }
+  result[result.size() - 1].pinning_mode = PinningMode::RECURSIVE;
```
Comment (supermassive) [inline components/ipfs/pin/ipfs_local_pin_service.cc:66]: nit: `result.back()`

**Review by supermassive — APPROVED** [review]

**Comment by srirambv** [comment]
Verification passed on

Brave | 1.52.17 Chromium: 112.0.5615.49 (Official Build) nightly (64-bit)
-- | --
Revision | `bd2a7bcb881c11e8cfe3078709382934e3916914-refs/branch-heads/5615@{#936}`
OS | Windows 11 Version 22H2 (Build 22621.1485)

- Verified steps from PR
- Verified NFT folder metadata is pinned directly
- Verified NFT metadata is pinned recursively 

https://user-images.githubusercontent.com/17010094/230020697-16c1f0ed-006b-40d3-8898-dddb61046126.mp4

**Pull request merged**